### PR TITLE
SignalR Sample

### PR DESCRIPTION
**Pull request by DovydasNavickas**
A sample project with basic ASOS and JwtBearerAuthentication for SignalR usage.

### CODE REVIEW

File: HelloSignalR/Connections/SimpleConnection.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using Microsoft.AspNet.Http;
+using Microsoft.Extensions.Logging;
+
+namespace HelloSignalR.Connections
+{
+    public class SimpleConnection : PersistentConnection
+    {
+        protected override async Task OnConnected(HttpRequest request, string connectionId)
+        {
+            await base.OnConnected(request, connectionId);
```
Comment (kevinchalet) [inline HelloSignalR/Connections/SimpleConnection.cs]: FYI, calling `base.OnConnected` is not useful as it just returns a completed task: https://github.com/aspnet/SignalR-Server/blob/dev/src/Microsoft.AspNet.SignalR.Server/PersistentConnection.cs#L431

File: HelloSignalR/Connections/SimpleConnection.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            var identity = request.HttpContext.User.Identity;
+            var authenticatedOrNot = (identity.IsAuthenticated ? "Authenticated" : "Unauthenticated");
```
Comment (kevinchalet) [inline HelloSignalR/Connections/SimpleConnection.cs]: `authenticatedOrNot` -> `status`? :smile:

File: HelloSignalR/Project_Readme.html [hunk 3]
```diff
@@ -0,0 +1,203 @@
+﻿<!DOCTYPE html>
```
Comment (kevinchalet) [inline HelloSignalR/Project_Readme.html]: Question: is it worth keeping this default file?

File: HelloSignalR/Providers/AuthenticationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+using AspNet.Security.OpenIdConnect.Extensions;
+
+namespace HelloSignalR.Providers
+{
+    public class AuthenticationProvider : OpenIdConnectServerProvider
+    {
+        public override Task ValidateClientAuthentication(ValidateClientAuthenticationContext context)
+        {
+            context.Skipped();
+            return base.ValidateClientAuthentication(context);
```
Comment (kevinchalet) [inline HelloSignalR/Providers/AuthenticationProvider.cs]: FYI, you don't need to call the base method, unless you also want to call the `OnValidateClientAuthentication` delegate.

File: HelloSignalR/Providers/AuthenticationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+using Microsoft.AspNet.Authentication.OpenIdConnect;
+using AspNet.Security.OpenIdConnect.Extensions;
+
+namespace HelloSignalR.Providers
+{
+    public class AuthenticationProvider : OpenIdConnectServerProvider
+    {
+        public override Task ValidateClientAuthentication(ValidateClientAuthenticationContext context)
+        {
+            context.Skipped();
```
Comment (kevinchalet) [inline HelloSignalR/Providers/AuthenticationProvider.cs]: Please consider adding a check here to ensure the client_id is the intended one. A comment indicating why `Skipped` is called might also be worth it.

File: HelloSignalR/Providers/AuthenticationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            return base.ValidateClientAuthentication(context);
+        }
+        public override Task GrantResourceOwnerCredentials(GrantResourceOwnerCredentialsContext context)
+        {
+            var user = new { Id = "users-123", UserName = "AspNet", Password = "contrib" };
+            if (context.UserName != user.UserName || context.Password != user.Password)
+            {
+                context.Rejected("Invalid username or password.");
+                return Task.FromResult(Task.FromResult(true));
```
Comment (kevinchalet) [inline HelloSignalR/Providers/AuthenticationProvider.cs]: lol, too many tasks here :smile: 

Just use `return Task.FromResult(0);` or `return Task.FromResult<object>(true);` :+1:

File: HelloSignalR/Providers/AuthenticationProvider.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            }
+
+            var identity = new ClaimsIdentity(OpenIdConnectDefaults.AuthenticationScheme);
+            identity.AddClaim(ClaimTypes.NameIdentifier, user.Id, "id_token token");
+            identity.AddClaim(ClaimTypes.Name, user.UserName, "id_token token");
+
+            var principal = new ClaimsPrincipal(identity);
+            context.Validated(principal);
+
+            return Task.FromResult(Task.FromResult(true));
```
Comment (kevinchalet) [inline HelloSignalR/Providers/AuthenticationProvider.cs]: Same remark :smile:

File: HelloSignalR/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        {
+            // Add the platform handler to the request pipeline.
+            app.UseIISPlatformHandler();
+
+            if (env.IsDevelopment())
+            {
+                // In case any errors occur.
+                app.UseDeveloperExceptionPage();
+
+                if (env.IsDevelopment())
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: Nice redundant check :smile:

File: HelloSignalR/Startup.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                {
+                    loggerFactory.MinimumLevel = LogLevel.Information;
+                    loggerFactory.AddConsole();
+                    loggerFactory.AddDebug();
+                }
+            }
+
+            logger.LogInformation($"Environment: {env.EnvironmentName}");
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: The hosting layer already logs the environment name, doing it manually is not necessary (and redundant).

File: HelloSignalR/Startup.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            app.UseOpenIdConnectServer(config =>
+            {
+                // Your Auhentication server.
+                config.Provider = new AuthenticationProvider();
+                config.Issuer = new Uri(hostname);
+
+                if (env.IsDevelopment())
+                {
+                    // For not requiring HTTPS in localhost.
+                    config.AllowInsecureHttp = true;
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: `AllowInsecureHttp` is automatically set to `true` by `UseOpenIdConnectServer` for development and testing environments.

File: HelloSignalR/Startup.cs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+                    config.ApplicationCanDisplayErrors = true;
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: That's actually a lie, you don't have anything to display the OIDC errors (well, you don't even have an authorization endpoint :smile:). Consider removing it.

File: HelloSignalR/Startup.cs [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                // Your audience that has to be allowed by your Authority server.
+                options.Audience = hostname;
+
+                // Authority that issued your token (i.e. your identity server).
+                options.Authority = hostname;
+
+                options.Events = new JwtBearerEvents
+                {
+                    // For SignalR connections default Authorization header does not work,
+                    // Because there are no Headers in WebSockets specification
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: Actually, that's not totally true: you **can** send HTTP headers with the initial WS handshake (on the SignalR client for .NET, there's a `Headers` property you can use for that). It's just that the WS JS API doesn't expose anything for that, which is a real shame.

File: HelloSignalR/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+                    // This is why we have to implement our own way to retrieve token
+                    // Which can be as easy as adding it to query string
+                    OnReceivingToken = context =>
+                    {
+                        string token = null;
+                        // TODO: Maybe these constants are defined somewhere in framework already?
+                        const string AuthorizationKey = "Authorization";
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: FYI, you can use `HeaderNames` for that.

File: HelloSignalR/Startup.cs [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+                        const string BearerKey = "Bearer";
+
+                        // First, try to extract token from 'Authorization: Bearer {token}' header
+                        if (context.Request.Headers.ContainsKey(AuthorizationKey))
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: The JWT bearer middleware already does that for you if you leave `context.Token` to null. Consider removing this whole block :smile:

File: HelloSignalR/Startup.cs [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+                            var start = $"{BearerKey} ";
+                            if (bearer.StartsWith(start) && bearer.Length > start.Length)
+                            {
+                                token = bearer.Substring(start.Length);
+                            }
+                        }
+
+                        // If the token is not there, look at '?token={token}' query string value
+                        token = token ?? context.Request.Query["token"];
```
Comment (kevinchalet) [inline HelloSignalR/Startup.cs]: Consider using `access_token` to match the standard name.

**Comment by kevinchalet** [comment]
Really good job, Dovydas! :+1: :clap: 

I'm really concerned about the security aspects of flowing the access token in the query string, which could potentially lead to session fixation-like attacks. We should maybe consider updating the `OnReceivingToken` event to ensure we only retrieve the token from the query string if it the request path corresponds to the SignalR path?

**Comment by kevinchalet** [comment]
nit: consider adding a `samples` directory at the root of the solution :smile:

File: HelloSignalR/Startup.cs (hunk 12 above)
Comment (DovydasNavickas) [inline HelloSignalR/Startup.cs]: Well, WS specification does not require Headers implementation, which is why WS JS API does not implement it. Why .NET client implements that is another question.

File: HelloSignalR/Startup.cs (hunk 8 above)
Comment (DovydasNavickas) [inline HelloSignalR/Startup.cs]: Ehm... Yeah.. :D Removed.

File: HelloSignalR/Startup.cs (hunk 9 above)
Comment (DovydasNavickas) [inline HelloSignalR/Startup.cs]: Didn't notice that... Thanks!

File: HelloSignalR/Startup.cs (hunk 14 above)
Comment (DovydasNavickas) [inline HelloSignalR/Startup.cs]: Removed :)

File: HelloSignalR/Startup.cs (hunk 15 above)
Comment (DovydasNavickas) [inline HelloSignalR/Startup.cs]: Updated.
https://github.com/QuatroDevOfficial/AspNet.Security.OpenIdConnect.Samples/commit/6b1282ce01cd0288e8bb5d5fdfe571baddbb688d

**Comment by DovydasNavickas** [comment]
I updated sample according to all comments except for moving to `samples` directory and checking if it's SignalR path in `OnReceivingToken`, because you can set custom path for SignalR. I'm gonna look if there is a way to figure out if the path is set to serve SignalR.

As for `samples` directory, I'm not sure that is really needed because the repository is already dedicated for samples anyway.

**Comment by kevinchalet** [comment]
> As for samples directory, I'm not sure that is really needed because the repository is already dedicated for samples anyway.

ATM, it's not really needed since we have only one example :smile: 
But in the future, we'll probably want to avoid having too many folders in the root.

**Comment by DovydasNavickas** [comment]
As it does not make much of a difference and in case we are planning to move projects either way, I moved it already.
If we leave the checking for SignalR connection for a bit later time, I think you can merge the PR.

File: samples/HelloSignalR/Startup.cs [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+                options.Events = new JwtBearerEvents
+                {
+                    // For SignalR connections default Authorization header does not work,
+                    // Because there are no Headers in WebSockets specification
+                    // This is why we have to implement our own way to retrieve token
+                    // Which can be as easy as adding it to query string
+                    OnReceivingToken = context =>
+                    {
+                        // If the token is not there, look at '?token={token}' query string value
+                        context.Token = context.Token ?? context.Request.Query["access_token"];
```
Comment (kevinchalet) [inline samples/HelloSignalR/Startup.cs]: The `context.Token ??` part is not necessary, it will always be null anyway (the default logic is applied after the OnReceivingToken event completes).

File: samples/HelloSignalR/Startup.cs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+                 {
+                     await next();
+                 }
+                 else {
+                     await context.Response.WriteAsync($"Username: {identity.Name}");
+                 }
+             });
+
+            // Configure the HTTP request pipeline for our front-end files.
+            app.UseDefaultFiles();
```
Comment (kevinchalet) [inline samples/HelloSignalR/Startup.cs]: Consider moving these 2 middleware just before the OIDC server :smile:

File: samples/HelloSignalR/wwwroot/index.html [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    <meta charset="utf-8" />
+    <script src="https://code.jquery.com/jquery-2.1.4.js" type="text/javascript"></script>
+    <script src="Scripts/jquery.signalR-2.2.0.js" type="text/javascript"></script>
+    <link href="styles/sample.css" rel="stylesheet" />
+    <title>ASOS SignalR Sample</title>
+    <script type="text/javascript">
+        'use strict';
+        $(document).ready(function () {
+            let connectionBase = "http://localhost:5000";
+            let connectionPath = `${connectionBase}/signalr`;
```
Comment (kevinchalet) [inline samples/HelloSignalR/wwwroot/index.html]: Looks like some new ECMAScript goodness is heavily used on this file :smile: 
Can we safely assume it will work on most browsers or is it worth using a transpiler thingy?

File: samples/HelloSignalR/wwwroot/index.html [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+            let connectToSignalR = (token) => {
+                let connection = $.connection(connectionPath);
+                let log = data => console.log(data);
+
+                connection.qs = `access_token=${token}`;
+
+                connection.received(data => {
+                    log(`From server: ${data}`);
```
Comment (kevinchalet) [inline samples/HelloSignalR/wwwroot/index.html]: That would be great if - instead of logging - you could just update the UI :smile:

**Comment by kevinchalet** [comment]
> I updated sample according to all comments except for moving to samples directory and checking if it's SignalR path in OnReceivingToken, because you can set custom path for SignalR. I'm gonna look if there is a way to figure out if the path is set to serve SignalR.

Looks like you configured the SignalR middleware to bind the connection to /signalr. Maybe you could simply use that?

File: AspNet.Security.OpenIdConnect.Samples.sln [hunk 20]
```diff
@@ -1,9 +1,27 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 14
-VisualStudioVersion = 14.0.23107.0
+VisualStudioVersion = 14.0.24627.0
 MinimumVisualStudioVersion = 10.0.40219.1
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "HelloSignalR", "HelloSignalR\HelloSignalR.xproj", "{BB7CE31F-BC7C-4F8B-8F05-19024ABD064D}"
```
Comment (kevinchalet) [inline AspNet.Security.OpenIdConnect.Samples.sln]: Don't forget to update the path or VS won't be able to load your project :smile:

File: samples/HelloSignalR/Startup.cs (hunk 17 above)
Comment (DovydasNavickas) [inline samples/HelloSignalR/Startup.cs]: Should I even leave it here? It was for JwtBearerAuthentication easy testing.

File: samples/HelloSignalR/wwwroot/index.html (hunk 19 above)
Comment (DovydasNavickas) [inline samples/HelloSignalR/wwwroot/index.html]: I know, that's why I used `log` as a proxy function. I will update the sample to update the UI later on. Want to have a stable first version first, because updating UI is just a fancy thing, not a crucial one :)

**Comment by DovydasNavickas** [comment]
Yeah, I could use '/signalr'. But there can be multiple SignalR endpoints, which makes it a bit tricky then.
If I one looks at the sample and hardcodes '/signalr', later on adding more SignalR endpoints makes a necessity to add them into `OnReceivingToken` method.

File: samples/HelloSignalR/Startup.cs (hunk 17 above)
Comment (DovydasNavickas) [inline samples/HelloSignalR/Startup.cs]: I think I will leave it here and add readme.md for project and explain steps how one can use the sample to understand how it works. And then this little middleware starts to make sense.

File: samples/HelloSignalR/wwwroot/index.html (hunk 18 above)
Comment (DovydasNavickas) [inline samples/HelloSignalR/wwwroot/index.html]: Good note. I've tested this on Chrome and MS Edge and it works like a charm.
Firefox and IE have some different opinions about using some of the things...
I will add transpiler.

**Comment by DovydasNavickas** [comment]
I've manually transpiled ES6 code to ES5 and now it works on all browsers OOTB.
Maybe when I add UI manipulations I'll go back to ES6 with transpiler, but I think there is an unnecessary burden for this sample with console to have ES6 with transpiler, because most of developers will have backend knowledge, but not necessarily have frontend knowledge about ES6 and transpilers.

**Comment by kevinchalet** [comment]
Looks fine :+1: 

One minor remark: the other repositories use K&R bracing style. Do you want me to update your PR before merging it or is it something you can easily do?

Also, don't hesitate to squash your commits :smile:

File: samples/HelloSignalR/Startup.cs [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+                 {
+                     await next();
+                 }
+                 else {
+                     await context.Response.WriteAsync($"Username: {identity.Name}");
+                 }
+             });
+
+            // Add token issuing middleware.
+            app.UseOpenIdConnectServer(config =>
```
Comment (kevinchalet) [inline samples/HelloSignalR/Startup.cs]: Hum, the order is weird now. Why did you change it?

File: samples/HelloSignalR/Startup.cs [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+            {
+                // Your Auhentication server.
+                config.Provider = new AuthenticationProvider();
+                config.Issuer = new Uri(hostname);
+            });
+
+            // Configure the HTTP request pipeline for our front-end files.
+            app.UseDefaultFiles();
```
Comment (kevinchalet) [inline samples/HelloSignalR/Startup.cs]: These middleware come too late, put them very early in the pipeline (after UseDeveloperExceptionPage).

**Comment by DovydasNavickas** [comment]
Updated middlewares order and brace styling to K&R, though I like Allman style more myself (most of AspNet stack uses it).
If everything is good to go, I will squash the commits then.

File: samples/HelloSignalR/Startup.cs (hunk 21 above)
Comment (kevinchalet) [inline samples/HelloSignalR/Startup.cs]: Consider moving that after the JWT middleware. With the current order, the OIDC server won't be reached if the previous inline middleware doesn't call next().

**Comment by DovydasNavickas** [comment]
https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Samples/pull/1#discussion_r45983059
What about now? It has to be good or I'm missing something here :D

File: samples/HelloSignalR/Providers/AuthenticationProvider.cs [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Authentication.OpenIdConnect;
+using AspNet.Security.OpenIdConnect.Extensions;
+
+namespace HelloSignalR.Providers {
+    public class AuthenticationProvider : OpenIdConnectServerProvider {
+        public override Task ValidateClientAuthentication(ValidateClientAuthenticationContext context) {
+            if (context.ClientId == "AspNetContribSample") {
+                context.Validated();
```
Comment (kevinchalet) [inline samples/HelloSignalR/Providers/AuthenticationProvider.cs]: Curious: why did you change it to `Validated()`? `Skipped()` was more appropriate.

**Comment by kevinchalet** [comment]
Minor one last remark, it's fine. Squash it and I'll merge it :+1:

File: samples/HelloSignalR/Providers/AuthenticationProvider.cs (hunk 23 above)
Comment (DovydasNavickas) [inline samples/HelloSignalR/Providers/AuthenticationProvider.cs]: Hmm... May be because there is no documentation on this and I literally am guessing which one fits best.
It also means that we really need to work on documentation at least part of the time :)

Comment (DovydasNavickas) [inline samples/HelloSignalR/Providers/AuthenticationProvider.cs]: Or at least great samples with readme.md :D

Comment (kevinchalet) [inline samples/HelloSignalR/Providers/AuthenticationProvider.cs]: Agreed :smile: 

In a nutshell, `Skipped` (or `Skip` in the nightly builds) is like `Validated` but for clients you can't trust (typically, JS or mobile clients): the request is not rejected, but ASOS is informed that the caller cannot be trusted and that the token is not confidential.

Comment (DovydasNavickas) [inline samples/HelloSignalR/Providers/AuthenticationProvider.cs]: Actually, good samples are a great start for documentation, because then we can write it more easily by simply explaining some code and referring to working samples.

**Comment by kevinchalet** [comment]
Merged, thanks! https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Samples/commit/06be66cd96d568ebc0db3f99370af61246282b6a :tada:
